### ISSUE TITLE
[SAC] Potential bug in temperature learning

### ISSUE THREAD

**Issue by johannespitz** [DLR-RM/stable-baselines3#36]
https://github.com/DLR-RM/stable-baselines3/blob/78e8d405d7bf6186c8529ed26967cb17ccbe420c/stable_baselines3/sac/sac.py#L184

In the [paper](https://arxiv.org/pdf/1812.05905.pdf) they compute the loss as `-alpha * (logp + entropy_target)` but here you implemented `-log(alpha) * (logp + entorpy_target)` if I am not mistaken.

I guess it should look like:
```
            ent_coef_loss = None
            if self.ent_coef_optimizer is not None:
                # Important: detach the variable from the graph
                # so we don't change it with other losses
                # see https://github.com/rail-berkeley/softlearning/issues/60
                ent_coef = th.exp(self.log_ent_coef) # No need to detach here anyways
                ent_coef_loss = -(ent_coef * (log_prob.detach() + self.target_entropy)).mean()
                ent_coef_losses.append(ent_coef_loss.item())
            else:
                ent_coef = self.ent_coef_tensor
```
Of course you could detach `ent_coef` afterwards in both cases, to maybe save time when computing `actor_loss` and doing the respective `.backward()`.

**Comment by johannespitz** [comment]
Interesting:
https://github.com/hill-a/stable-baselines/blob/c4c31cb5687800cf102054f0935aa09e7545159f/stable_baselines/sac/sac.py#L231
https://github.com/rail-berkeley/softlearning/blob/05daa5524ae1a76b70b8a8a29a0f5f824d401484/softlearning/algorithms/sac.py#L250

**Comment by araffin** [comment]
Hello,
Thanks for pointing that out.
There are several issues about that:
- https://github.com/rail-berkeley/softlearning/issues/37
- https://github.com/hill-a/stable-baselines/issues/801
- https://github.com/hill-a/stable-baselines/issues/828

The answer is yes, you are right, in practice, it does not change the results (the original implementation was doing that before).

**Comment by johannespitz** [comment]
Ah somewhat related I guess:
https://github.com/DLR-RM/stable-baselines3/blob/91adefdb4b9aff2e0c5f2dfad47877a74bcc0de6/stable_baselines3/common/distributions.py#L599

The latest version of pytorch finally has a tanh transform as well.
https://pytorch.org/docs/stable/distributions.html#torch.distributions.transforms.TanhTransform
Not sure how that relates exactly to the numerical problems you mentioned in your edit.

I guess in the end the only goal of this loss is to control the entropy but I am still surprised that it works like this.

**Comment by araffin** [comment]
>The latest version of pytorch finally has a tanh transform as well.

thanks, I did not know that =)

>but I am still surprised that it works like this.

Look at the first link, this should be equivalent

**Comment by johannespitz** [comment]
All right assuming that we constrain `alpha >= 0` in the last line but not in the first two lines then the `min` is the same. However, you go on and use `alpha` to compute the other losses. Meaning we  want to know if the `argmin` is the same.
```
min E[-alpha * log_pi - alpha * H]
= min E[-alpha * (log_pi + H)]
= min E[-log alpha * (log_pi + H)]
```
I believe that what's really going on here is that the gradient of `alpha` or some `alpha_logit` has the same sign depending on whether `log_pi + H` is positive or negative and then Adam does its usual magic.
Anyways, personally I would prefer to use the originally derived formula if it doesn't degrade the performance...
